src/amd_ipmi.c

```diff
@@ -0,0 +1,435 @@
+ * /sys/devices/pci0000:e0/0000:e0:03.3/0000:e3:00.0
+ *                                      ^^^^^^^^^^
+ * We need to retrieve the address from the path (indicated above)
+ * then use it to find the corresponding address for a slot in
+ * /sys/bus/pci_slots to determine the icorrect port for the NVMe device.
+ */
+static int _get_ipmi_nvme_port(char *path)
+{
+	int rc;
+	char *p, *f;
+	struct list dir;
+	const char *dir_path;
+	char *port_name;
+	int port = -1;
+
+	p = strrchr(path, '/');
+	p++;
+
+	/* p now points to the address, remove the bits after the '.' */
+	f = strchr(p, '.');
+	*f = '\0';
+
+	rc = scan_dir("/sys/bus/pci/slots", &dir);
+	if (rc)
+		return -1;
+
+	list_for_each(&dir, dir_path) {
+		port_name = get_text(dir_path, "address");
+		if (!strcmp(port_name, p)) {
+			char *dname = strrchr(dir_path, '/');
+
+			dname++;
+			port = atoi(dname);
+			break;
+		}
+	}
+
+	list_erase(&dir);
+
+	/* Some platfroms require an adjustment to the port value based
+	 * on how they are numbered by the BIOS.
+	 */
+	switch (amd_platform) {
+	case AMD_PLATFORM_DAYTONA_X:
+		port -= 2;
+		break;
+	case AMD_PLATFORM_ETHANOL_X:
+		port -= 7;
+		break;
+	default:
+		break;
+	}
+
+	/* Validate port. Some BIOSes provide provide port values that are
+	 * not valid.
+	 */
+	if ((port < 0) || (port > 24)) {
+		log_error("Invalid NVMe physical port %d\n", port);
+		port = -1;
+	}
+
+	return port;
+}
+
+static int _get_ipmi_sata_port(const char *start_path)
+{
+	int port;
+	char *p, *t;
+	char path[PATH_MAX];
+
+	strncpy(path, start_path, PATH_MAX);
+	path[PATH_MAX - 1] = 0;
+	t = p = strstr(path, "ata");
+
+	if (!p)
+		return -1;
+
+	/* terminate the path after the ataXX/ part */
+	p = strchr(p, '/');
+	if (!p)
+		return -1;
+	*p = '\0';
+
+	/* skip past 'ata' to get the ata port number */
+	t += 3;
+	port = strtoul(t, NULL, 10);
+
+	return port;
+}
+
+static int _get_amd_ipmi_drive(const char *start_path,
+			       struct amd_drive *drive)
+{
+	int found;
+	char path[PATH_MAX];
+
+	found = _find_file_path(start_path, "nvme", path, PATH_MAX);
+	if (found) {
+		drive->port = _get_ipmi_nvme_port(path);
+		if (drive->port < 0) {
+			log_error("Could not retrieve port number\n");
+			return -1;
+		}
+
+		drive->drive_bay = 1 << (drive->port - 1);
+		drive->dev = AMD_NVME_DEVICE;
+	} else {
+		int shift;
+
+		drive->port = _get_ipmi_sata_port(start_path);
+		if (drive->port < 0) {
+			log_error("Could not retrieve port number\n");
+			return -1;
+		}
+
+		/* IPMI control is handled through the MG9098 chips on
+		 * the platform, where each MG9098 chip can control up
+		 * to 8 drives. Since we can have multiple MG9098 chips,
+		 * we need the drive bay relative to the set of 8 controlled
+		 * by the MG9098 chip.
+		 */
+		shift = drive->port - 1;
+		if (shift >= 8)
+			shift %= 8;
+
+		drive->drive_bay = 1 << shift;
+		drive->dev = AMD_SATA_DEVICE;
+	}
+
+	if (drive->port == -1) {
+		log_info("Couldn't find drive for %s\n", start_path);
```

Comment:
This code is unreachable. For drive->port == -1 function will exit in line 157 or 168.